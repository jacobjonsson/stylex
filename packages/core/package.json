{
  "name": "@jacobjonsson/stylex",
  "version": "0.0.1",
  "description": "Compile time css in js",
  "main": "dist/index.js",
  "repository": "https://github.com/jacobjonsson/stylex",
  "author": "Jacob Jonsson <jacobjjonsson@gmail.com>",
  "license": "MIT",
  "private": false,
  "scripts": {
    "build": "yarn run rollup -c",
    "test": "yarn run jest"
  },
  "dependencies": {
    "@babel/plugin-syntax-jsx": "^7.8.3",
    "css": "^2.2.4",
    "find-yarn-workspace-root": "1.2.1",
    "loader-utils": "2.0.0",
    "mkdirp": "1.0.4",
    "stylis": "^3.5.4"
  }
}
