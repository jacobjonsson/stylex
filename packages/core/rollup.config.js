import { rmdirSync } from 'fs';
import { join } from 'path';
import babel from 'rollup-plugin-babel';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';

rmdirSync(join(__dirname, 'dist'), { recursive: true });

export default [
    {
        input: {
            babel: 'src/babel.js',
            loader: 'src/loader.js',
        },
        external: [
            '@babel/plugin-syntax-jsx',
            'css',
            'stylis',
            'loader-utils',
            '@babel/core',
            'path',
            'fs',
            'mkdirp',
        ],
        output: {
            dir: 'dist',
            format: 'cjs',
        },
        plugins: [babel({ exclude: 'node_modules/**' }), resolve(), commonjs()],
    },
    {
        input: 'src/index.js',
        output: {
            file: 'dist/index.js',
            format: 'esm',
        },
        plugins: [babel({ exclude: 'node_modules/**' }), resolve()],
    },
];
