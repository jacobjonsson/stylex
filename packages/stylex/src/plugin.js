import jsx from '@babel/plugin-syntax-jsx';
import generate from 'babel-generator';
import loadConfig from './utils/loadConfig';

export function createBabelPlugin(options) {
    const { store } = options;

    return function babelPlugin(babel) {
        const theme = loadConfig().theme;

        return {
            name: 'stylex',
            inherits: jsx,
            visitor: {
                TaggedTemplateExpression(path) {
                    let t = babel.types;
                    let { quasi, tag } = path.node;

                    if (tag.name !== 'stylex') {
                        return;
                    }

                    let cssText = '';
                    let expressions = path.get('quasi').get('expressions');

                    quasi.quasis.forEach((el, i) => {
                        cssText += el.value.cooked;
                        let expression = expressions[i];

                        if (t.isArrowFunctionExpression(expression)) {
                            const { code } = generate(expression.node);

                            // Execute the stringified function from the css
                            const value = new Function(`return ${code}`)()(theme);
                            cssText += value;
                        }
                    });

                    const className = store.create(cssText);
                    path.replaceWith(t.stringLiteral(className));
                },
            },
        };
    };
}
