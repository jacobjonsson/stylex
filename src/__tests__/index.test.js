import path from "path";
import { transformFileSync } from "@babel/core";
import plugin from "../babel";

function transpile(file) {
  return transformFileSync(path.resolve(__dirname, file), {
    plugins: [plugin],
    babelrc: false,
  });
}

it.each([["simple.js"], ["media-queries.js"], ["nested-selectors.js"]])("%s", (file) => {
  const { code, metadata } = transpile(`./fixtures/${file}`);
  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});
