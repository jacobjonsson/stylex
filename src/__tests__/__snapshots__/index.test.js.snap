// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media-queries.js 1`] = `
"import stylex from \\"stylex\\";
const styles = {
  root: \\"xadsd7g x1lyv7vr\\"
};
export function Component() {
  return <div className={stylex.resolve(styles.root)} />;
}"
`;

exports[`media-queries.js 2`] = `
Object {
  "stylex": ".xadsd7g {font-size: 12;}
@media (min-width: 768px): {.x1lyv7vr {font-size: 14;}}",
}
`;

exports[`nested-selectors.js 1`] = `
"import stylex from \\"stylex\\";
const styles = {
  root: \\"xadsd7g xmsj4j\\"
};
export function Component() {
  return <div className={stylex.resolve(styles.root)} />;
}"
`;

exports[`nested-selectors.js 2`] = `
Object {
  "stylex": ".xadsd7g {font-size: 12;}
@media (min-width: 768px): {.x1lyv7vr {font-size: 14;}}
.xmsj4j > p {color: white;}",
}
`;

exports[`simple.js 1`] = `
"import stylex from \\"stylex\\";
const styles = {
  root: \\"xadsd7g\\"
};
export function Component() {
  return <div className={stylex.resolve(styles.root)} />;
}"
`;

exports[`simple.js 2`] = `
Object {
  "stylex": ".xadsd7g {font-size: 12;}",
}
`;
